### PR TITLE
Add kotlin transformer

### PR DESCRIPTION
**Pull request by abdullahtariq1171**
# ↪️ Pull Request

Closes https://github.com/parcel-bundler/parcel/issues/3356

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/transformers/kotlin/src/KotlinTransformer.js [hunk 1]
```diff
@@ -1,11 +1,57 @@
 // @flow
 
 import {Transformer} from '@parcel/plugin';
+import fs from '@parcel/fs';
```
Comment (Taym95) [inline packages/transformers/kotlin/src/KotlinTransformer.js]: I think you don't need `fs` from `@parcel/fs` you can use `asset.fs`.

File: packages/transformers/kotlin/src/KotlinTransformer.js [hunk 2]
```diff
@@ -1,11 +1,57 @@
...
+      asset.filePath
+    );
+
+    let id = Math.random()
+      .toString(36)
+      .slice(3);
+    let dir = path.join(os.tmpdir(), id);
+    let filename = path.join(dir, id + '.js');
+
+    await fs.mkdirp(dir);
```
Comment (Taym95) [inline packages/transformers/kotlin/src/KotlinTransformer.js]: You can use this `asset.fs.mkdirp(dir)`.

File: packages/transformers/kotlin/src/KotlinTransformer.js [hunk 3]
```diff
@@ -1,11 +1,57 @@
...
+      output: filename,
+      sources: [asset.filePath],
+      moduleKind: 'commonjs',
+      noStdlib: false,
+      metaInfo: true,
+      sourceMaps: options.sourceMaps
+    });
+
+    let code = await fs.readFile(filename, 'utf8');
+
```
Comment (Taym95) [inline packages/transformers/kotlin/src/KotlinTransformer.js]: Here also you can use this `asset.fs.readFile(filename, 'utf8')`.

File: packages/transformers/kotlin/src/KotlinTransformer.js [hunk 4]
```diff
@@ -1,11 +1,57 @@
...
+    if (options.sourceMaps) {
+      let sourceMap = await fs.readFile(filename + '.map', 'utf8');
+
```
Comment (Taym95) [inline packages/transformers/kotlin/src/KotlinTransformer.js]: Same here you can use `asset.fs`.

File: packages/transformers/kotlin/src/KotlinTransformer.js [hunk 5]
```diff
@@ -1,11 +1,57 @@
...
+      asset.setMap(sourceMap);
+
+      // remove source map url
+      code = code.substring(0, code.lastIndexOf('//# sourceMappingURL'));
+    }
+
+    asset.setCode(code);
+
+    // delete temp directory
+    await fs.rimraf(dir);
```
Comment (Taym95) [inline packages/transformers/kotlin/src/KotlinTransformer.js]: Same here you can use `asset.fs`.

File: packages/transformers/kotlin/src/KotlinTransformer.js (hunk 1 above)
Comment (mischnic) [reply packages/transformers/kotlin/src/KotlinTransformer.js]: (Correct, you *have* to use `asset.fs`)

File: packages/core/integration-tests/test/kotlin.js [hunk 6]
```diff
@@ -14,7 +14,7 @@ describe.skip('kotlin', function() {
   it('should produce a basic kotlin bundle', async function() {
     let b = await bundle(__dirname + '/integration/kotlin/index.js');
 
-    await assertBundleTree(b, {
+    await assertBundles(b, {
```
Comment (abdullahtariq1171) [inline packages/core/integration-tests/test/kotlin.js]: test fails here with `TypeError: expectedBundles is not iterable`.

**Comment by abdullahtariq1171** [comment]
**Question:**

> Collecting all files referenced in the code so that Parcel watches them (not currently supported in Parcel 1)

How can we collect dependencies here? I see in some transformers we get ast which we visit and extract imports but here we don't parse the code instead we get generated output code and maps from `@jetbrains/kotlinc-js-api` (`kotlin-compiler` wrapper)

File: packages/core/integration-tests/test/kotlin.js (hunk 6 above)
Comment (mischnic) [reply packages/core/integration-tests/test/kotlin.js]: the second argument has to be an array, e.g.:
```js
    assertBundles(b, [
      {
        type: 'js',
        assets: ['index.js']
      }
    ]);
```

**Pull request closed without merging**